File: app/src/net/reichholf/dreamdroid/widget/helper/ClickItemTouchListener.java
package net.reichholf.dreamdroid.widget.helper;

import android.os.Build;
import androidx.core.view.GestureDetectorCompat;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.RecyclerView.OnItemTouchListener;
import android.view.GestureDetector.SimpleOnGestureListener;
import android.view.MotionEvent;
import android.view.View;

abstract class ClickItemTouchListener implements OnItemTouchListener {
	private final GestureDetectorCompat mGestureDetector;

	ClickItemTouchListener(RecyclerView hostView) {
		mGestureDetector = new GestureDetectorCompat(hostView.getContext(),
				new ItemClickGestureListener(hostView));
	}

	private boolean isAttachedToWindow(RecyclerView hostView) {
		if (Build.VERSION.SDK_INT >= 19) {
			return hostView.isAttachedToWindow();
		} else {
			return (hostView.getHandler() != null);
		}
	}

	private boolean hasAdapter(RecyclerView hostView) {
		return (hostView.getAdapter() != null);
	}

	@Override
	public boolean onInterceptTouchEvent(RecyclerView recyclerView, MotionEvent event) {
		if (!isAttachedToWindow(recyclerView) || !hasAdapter(recyclerView)) {
			return false;
		}

		mGestureDetector.onTouchEvent(event);
		return false;
	}

	@Override
	public void onTouchEvent(RecyclerView recyclerView, MotionEvent event) {
		// We can silently track tap and and long presses by silently
		// intercepting touch events in the host RecyclerView.
	}

	abstract boolean performItemClick(RecyclerView parent, View view, int position, long id);
	abstract boolean performItemLongClick(RecyclerView parent, View view, int position, long id);

	private class ItemClickGestureListener extends SimpleOnGestureListener {
		private final RecyclerView mHostView;
		private View mTargetChild;

		public ItemClickGestureListener(RecyclerView hostView) {
			mHostView = hostView;
		}

		@Override
		public boolean onDown(MotionEvent event) {
			final int x = (int) event.getX();
			final int y = (int) event.getY();

			mTargetChild = mHostView.findChildViewUnder(x, y);
			return (mTargetChild != null);
		}

		@Override
		public void onShowPress(MotionEvent event) {
			if (mTargetChild != null) {
				mTargetChild.setPressed(true);
			}
		}

		@Override
		public boolean onSingleTapUp(MotionEvent event) {
			boolean handled = false;

			if (mTargetChild != null) {
				mTargetChild.setPressed(false);

				final int position = mHostView.getChildAdapterPosition(mTargetChild);
				final long id = mHostView.getAdapter().getItemId(position);
				handled = performItemClick(mHostView, mTargetChild, position, id);

				mTargetChild = null;
			}

			return handled;
		}

		@Override
		public boolean onScroll(MotionEvent event, MotionEvent event2, float v, float v2) {
			if (mTargetChild != null) {
				mTargetChild.setPressed(false);
				mTargetChild = null;

				return true;
			}

			return false;
		}

		@Override
		public void onLongPress(MotionEvent event) {
			if (mTargetChild == null) {
				return;
			}

			final int position = mHostView.getChildAdapterPosition(mTargetChild);
			final long id = mHostView.getAdapter().getItemId(position);
			final boolean handled = performItemLongClick(mHostView, mTargetChild, position, id);

			if (handled) {
				mTargetChild.setPressed(false);
				mTargetChild = null;
			}
		}
	}
}

File: app/src/net/reichholf/dreamdroid/widget/helper/ClickItemTouchListener.kt
package net.reichholf.dreamdroid.widget.helper

import android.view.GestureDetector
import android.view.GestureDetector.SimpleOnGestureListener
import android.view.MotionEvent
import android.view.View
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.OnItemTouchListener

abstract class ClickItemTouchListener(hostView: RecyclerView) : OnItemTouchListener {
    private val gestureDetector: GestureDetector =
        GestureDetector(hostView.context, ItemClickGestureListener(hostView))

    private fun isAttachedToWindow(hostView: RecyclerView): Boolean = hostView.isAttachedToWindow

    private fun hasAdapter(hostView: RecyclerView): Boolean = hostView.adapter != null

    override fun onInterceptTouchEvent(recyclerView: RecyclerView, event: MotionEvent): Boolean {
        if (!isAttachedToWindow(recyclerView) || !hasAdapter(recyclerView)) {
            return false
        }
        gestureDetector.onTouchEvent(event)
        return false
    }

    override fun onTouchEvent(recyclerView: RecyclerView, event: MotionEvent) {
        // We can silently track tap and and long presses by silently
        // intercepting touch events in the host RecyclerView.
    }

    override fun onRequestDisallowInterceptTouchEvent(disallowIntercept: Boolean) {}

    abstract fun performItemClick(
        parent: RecyclerView,
        view: View,
        position: Int,
        id: Long
    ): Boolean
    abstract fun performItemLongClick(
        parent: RecyclerView,
        view: View,
        position: Int,
        id: Long
    ): Boolean

    private inner class ItemClickGestureListener(private val hostView: RecyclerView) :
        SimpleOnGestureListener() {
        private var targetChild: View? = null

        override fun onDown(event: MotionEvent): Boolean {
            val x = event.x.toInt()
            val y = event.y.toInt()
            targetChild = hostView.findChildViewUnder(x.toFloat(), y.toFloat())
            return targetChild != null
        }

        override fun onShowPress(event: MotionEvent) {
            targetChild?.isPressed = true
        }

        override fun onSingleTapUp(event: MotionEvent): Boolean {
            var handled = false
            val target = targetChild
            if (target != null) {
                target.isPressed = false
                val position = hostView.getChildAdapterPosition(target)
                val id = hostView.adapter!!.getItemId(position)
                handled = performItemClick(hostView, target, position, id)
                targetChild = null
            }
            return handled
        }

        override fun onScroll(
            event: MotionEvent?,
            event2: MotionEvent,
            v: Float,
            v2: Float
        ): Boolean {
            val target = targetChild
            if (target != null) {
                target.isPressed = false
                targetChild = null
                return true
            }
            return false
        }

        override fun onLongPress(event: MotionEvent) {
            val target = targetChild ?: return
            val position = hostView.getChildAdapterPosition(target)
            val id = hostView.adapter!!.getItemId(position)
            val handled = performItemLongClick(hostView, target, position, id)
            if (handled) {
                target.isPressed = false
                targetChild = null
            }
        }
    }
}
